tracks/haskell/exercises/nucleotide-count/mentoring.md

```diff
@@ -0,0 +1,163 @@
+### Reasonable solutions
+
+```haskell
+nucleotideCounts :: String -> Either String (Map Nucleotide Int)
+nucleotideCounts s
+  | all isNucleotide s = Right (counts `M.union` empty)
+  | otherwise          = Left s
+  where
+    nucleotides = map (read . return) s
+    counts = M.fromListWith (+) (zip nucleotides [1,1..])
+
+isNucleotide :: Char -> Bool
+isNucleotide = (`elem` "ACGT")
+
+empty :: Map Nucleotide Int
+empty = M.fromList [(A, 0), (C, 0), (G, 0), (T, 0)]
+```
+
+This solution handles errors in an initial guard.
+
+At its core, `M.fromListWith (+)` produces the result.  Because nucleotides
+with a count of zero must occur in the result, the initial map is populated
+with zeroes. Beware that *the order matters* for arguments to `M.union`!
+
+The error message is a little imprecise.
+
+```haskell
+import qualified Data.Map as M
+import           Data.Map (Map)
+
+data Nucleotide = A | C | G | T deriving (Eq, Ord, Show)
+
+nucleotideCounts :: String -> Either String (Map Nucleotide Int)
+nucleotideCounts s
+  | all isNucleotide s = Right $ M.fromList [ (A, count 'A')
+                                            , (C, count 'C')
+                                            , (G, count 'G')
+                                            , (T, count 'T') ]
+  | otherwise = Left $ "Invalid strand " ++ show s
+  where count c = length . filter (== c) $ s
+
+isNucleotide :: Char -> Bool
+isNucleotide = (`elem` "ACGT")
+```
+
+This solution also handles errors in an initial guard.
+
+Because GHC can fuse list combinators, four separate `length . filter (== c)`
+do not necessarily perform bad. Zeroes are not a special case here.
```

Comment:
This sentence seems a little fishy to me -- four separate `length . filter (== c)` calls "do not necessarily" perform bad?

It's true that GHC can fuse list combinators, but to the best of my knowledge this only works if you compose certain functions (at least catamorphisms). I'm not sure this claim is accurate when applied to four independent invocations of `count`.

Reply:
I'll have to run a benchmark then.

Sounds like you're right.

I recall this one not running nearly like four times for large input.